// messenger-api-helpers/lists.js
'use strict';
const AWS = require('aws-sdk');

const LIST_TABLE = process.env.LIST_TABLE;
const AWS_REGION = process.env.AWS_REGION;
const docClient = new AWS.DynamoDB.DocumentClient({region: AWS_REGION});

module.exports.getOwnedForUser = async (userId) => {
  const params = {
    TableName: LIST_TABLE,
    KeyConditionExpression: "#owner = :owner",
    ExpressionAttributeNames:{
        "#owner": "owner"
    },
    ExpressionAttributeValues: {
        ":owner": userId
    }
  };
  try {
    const data = await docClient.query(params).promise();
    return data.Items;
  } catch(error) {
    return [];
  }
};

module.exports.getSharedToUser = async (userId) => {
    const params = {
      TableName: LIST_TABLE,
      KeyConditionExpression: "#sharedwith = :sharedwith",
      ExpressionAttributeNames:{
          "#sharedwith": "sharedwith"
      },
      ExpressionAttributeValues: {
          ":sharedwith": userId
      },
      IndexName: "sharedwith-index"
    };
    try {
      const data = await docClient.query(params).promise();
      return data.Items;
    } catch(error) {
      return [];
    }
  };
